File: test/four/app/src/main/java/com/z3d/vo/SkinMesh.java
package com.z3d.vo;

import com.z3d.base.CallBackFun;
import com.z3d.base.MeshData;
import com.z3d.engine.ResCount;
import com.z3d.material.Material;
import com.z3d.material.MaterialBackFun;
import com.z3d.material.MaterialBaseParam;
import com.z3d.program.MaterialAnimShader;
import com.z3d.scene.Scene3D;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class SkinMesh extends ResCount {

    public List<MeshData> meshAry ;
    public List<Matrix3D> bindPosMatrixAry;
    public List<Matrix3D> bindPosInvertMatrixAry;
    public HashMap<String, BoneSocketData> boneSocketDic;
    public float fileScale ;
    public float tittleHeight ;
    public Vector2D hitBox ;
    public int type ;
    public HashMap animDic ;
    public boolean ready ;
    public List<String>  animUrlAry ;
    public List<List<Float> >  lightData ;
    public List<Vector3D> hitPosItem;
    public HashMap allParticleDic;
    public String url;
    public boolean hasDestory ;
    public SkinMesh(Scene3D val)
    {
        super(val);
        this.animDic=new HashMap();
        this.meshAry=new ArrayList<>();
    }
    public void makeHitBoxItem()
    {

    }
    public void addMesh(MeshData $mesh)
    {
        $mesh.uid = this.meshAry.size();
        this.meshAry.add($mesh);
    }
    public  void  loadMaterial(){
        for (int i = 0; i < this.meshAry.size(); i++){
            this.loadByteMeshDataMaterial(meshAry.get(i),null);
        }
    }

    private void loadByteMeshDataMaterial(MeshData meshData, CallBackFun backFun) {

        scene3D.materialManager.getMaterialByte(meshData.materialUrl, new MaterialBackFun() {
            @Override
            public void Bfun(Material value) {
                meshData.material = value;
                meshData.materialParam = new MaterialBaseParam(scene3D);
                meshData.materialParam.setData(value, meshData.materialParamData);
            }
        },true, MaterialAnimShader.shaderNameStr, new MaterialAnimShader(scene3D));

    }

    public  void  setAction(List<String> actionAry,String roleUrl )
    {
        this.animUrlAry=new ArrayList<>();
        for (int i = 0; i < actionAry.size(); i++) {
            String name  = actionAry.get(i);
            String url = roleUrl +name;
            AnimData anim=  scene3D.animManager.getAnimDataImmediate(url);
            anim.processMesh(this);
            this.animDic.put(name,anim);
            this.animUrlAry.add(url);
        }
    }


}

File: h5gl/src/Pan3d/vo/SkinMesh.ts
module Pan3d {

    export class SkinMesh extends ResCount {
        public animUrlAry: Array<string>;
        public animDic: Object = new Object;
        public meshAry: Array<MeshData> = new Array;
        allParticleDic: Object;
        boneSocketDic: Object;
        url: string;
        lightData: number[][];
        ready: boolean;
        public addMesh($mesh: MeshData): void {
            $mesh.uid = this.meshAry.length;
            this.meshAry.push($mesh);
        }
        makeHitBoxItem() {
 
        }
        fileScale: any;
        tittleHeight: any;
        hitBox: Vector2D;

        public loadMaterial($fun: Function = null): void {
            for (var i: number = 0; i < this.meshAry.length; i++) {
                this.loadByteMeshDataMaterial(this.meshAry[i], $fun);
            }
        }

        private loadByteMeshDataMaterial($meshData: MeshData, $fun: Function = null): void {
            var url: string = this.scene3D.fileRoot + $meshData.materialUrl;
            url = url.replace("_byte.txt", ".txt")
            url = url.replace(".txt", "_byte.txt")

            this.scene3D.materialManager.getMaterialByte(url, ($material: Material) => {
                $meshData.material = $material;
                if ($material.usePbr) {
                    this.scene3D.meshDataManager.uploadPbrMesh($meshData, $material.useNormal);
                } else if ($material.lightProbe || $material.directLight) {
                    this.scene3D.meshDataManager.uploadPbrMesh($meshData, false);
                }

                if ($meshData.materialParamData) {
                    $meshData.materialParam = new MaterialBaseParam(this.scene3D);
                    $meshData.materialParam.setData($meshData.material, $meshData.materialParamData);
                }

                if ($fun) {
                    $fun($material);
                }
            }, null, true, MaterialAnimShader.MATERIAL_ANIM_SHADER, MaterialAnimShader);
        }

        public setAction(actionAry: Array<string>, roleUrl: string): void {
            this.animUrlAry = new Array;
            for (var i: number = 0; i < actionAry.length; i++) {
                var name: string = actionAry[i];
                var url: string = roleUrl + actionAry[i];
                var anim: AnimData = this.scene3D.animManager.getAnimDataImmediate(url);
                anim.processMesh(this);
                this.animDic[name] = anim;
                this.animUrlAry.push(url);
            }
        }

    }
}